packages/state-transition/test/perf/block/processWithdrawals.test.ts

```diff
@@ -0,0 +1,103 @@
+    {id: "bestcase", lowBalanceRatio: 0, blsCredentialRatio: 0},
+    // Normal case based on mainnet conditions: mainnet network conditions: 95% reward rate
+    {id: "normalcase - 100% withdrawals enabled", lowBalanceRatio: 0.05, blsCredentialRatio: 0},
+    {id: "normalcase - 70% withdrawals enabled", lowBalanceRatio: 0.05, blsCredentialRatio: 0.3},
+    {id: "normalcase - 40% withdrawals enabled", lowBalanceRatio: 0.05, blsCredentialRatio: 0.6},
+    {id: "normalcase - 10% withdrawals enabled", lowBalanceRatio: 0.05, blsCredentialRatio: 0.9},
+    // Degraded case
+    {id: "slighly degraded case - 90% withdrawals enabled", lowBalanceRatio: 0.1, blsCredentialRatio: 0.1},
+    {id: "degraded case - 90% withdrawals enabled", lowBalanceRatio: 0.3, blsCredentialRatio: 0.1},
+    {id: "highly degraded case - 90% withdrawals enabled", lowBalanceRatio: 0.4, blsCredentialRatio: 0.1},
+    // Worst case: All effective balance are updated
+    // NOTE: The maximum bad case will practically never happen and it's too slow.
+    // Use a 50% worst case since it's not that slow.
+    {id: "worstcase - lowBalance ", lowBalanceRatio: 1, blsCredentialRatio: 0},
+    {id: "worstcase - no withdrawals enabled ", lowBalanceRatio: 0, blsCredentialRatio: 1},
+  ];
+
+  // Provide flat `epochProcess.balances` + flat `epochProcess.validators`
+  // which will it update validators tree
+
+  for (const {id, lowBalanceRatio, blsCredentialRatio} of testCases) {
+    itBench<{state: CachedBeaconStateCapella}, {state: CachedBeaconStateCapella}>({
+      id: `capella getExpectedWithdrawals - ${vc} ${id}`,
+      yieldEventLoopAfterEach: true, // So SubTree(s)'s WeakRef can be garbage collected https://github.com/nodejs/node/issues/39902
+      minRuns: 5, // Worst case is very slow
+      before: () => getExpectedWithdrawalsTestData(vc, lowBalanceRatio, blsCredentialRatio),
+      beforeEach: ({state}) => ({state: state.clone()}),
+      fn: ({state}) => {
+        getExpectedWithdrawals(state);
+      },
+    });
+  }
+});
+
+/**
+ * Create a state that has `lowBalanceRatio` fraction (0,1) of validators with balance < max effective
+ * balance and `blsCredentialRatio` fraction (0,1) of withdrawal credentials not set for withdrawals
+ */
+function getExpectedWithdrawalsTestData(
+  vc: number,
+  lowBalanceRatio: number,
+  blsCredentialRatio: number
+): {
+  state: CachedBeaconStateCapella;
+} {
+  const stateTree = ssz.capella.BeaconState.defaultViewDU();
+  stateTree.slot = 1;
+
+  const activeValidator = ssz.phase0.Validator.toViewDU({
+    ...ssz.phase0.Validator.defaultValue(),
+    exitEpoch: Infinity,
+    withdrawableEpoch: Infinity,
+    // Set current effective balance to max
+    effectiveBalance: 32e9,
+  });
+
+  const balances: number[] = [];
+  for (let i = 0; i < vc; i++) {
+    // Math.random() supposedly gives a uniform distribution between 0 & 1
+    const balance = lowBalanceRatio > Math.random() ? 30e9 : 33e9;
```

Comment:
Performance tests should be deterministic. It shouldn't matter the pattern of which specific indexes have low balance. You could do something like 
```ts
const hasLowBalance = i % Math.round(1 / lowBalanceRatio) === 0
```

Reply:
makes sense, will update